src/easier68k/core/enum/ea_mode.py

```diff
@@ -0,0 +1,145 @@
+"""
+Represents an effective addressing mode
+and methods associated with it
+"""
+from src.easier68k.core.util.parsing import parse_literal
+
+
+class EAMode:
+    # Enum Values
+    # Error value
+    ERR = -1
+
+    # Data register direct
+    DRD = 0
+
+    # Address register direct
+    ARD = 1
+
+    # Address register indirect
+    ARI = 2
+
+    # Address register indirect + post increment
+    ARIPI = 3
+
+    # Address register indirect + pre decrement
+    ARIPD = 4
+
+    # Immediate
+    IMM = 5
+
+    # Absolute long address
+    ALA = 6
+
+    # Absolute word address
+    AWA = 7
+
+    # Instance Values
+    # Which mode this represents
+    _mode = ERR
+
+    # Data value (could be register number, immediate data, or absolute address
+    _data = -1
+
+    def __init__(self, mode=ERR, data=-1):
+        self._mode = mode
+        self._data = data
+
+    def __str__(self):
+        return "Mode: {}, Data: {}".format(self._mode, self._data)
+
+    @staticmethod
+    def parse_ea(addr: str):
+        """
+        Parses an effective addressing mode (such as D0, (A1), #$01)
+
+        >>> EAMode.parse_ea('D')
+        Traceback (most recent call last):
+        ...
+        AssertionError
+
+        >>> str(EAMode.parse_ea('D3'))
+        'Mode: 0, Data: 3'
+
+        >>> str(EAMode.parse_ea('A6'))
+        'Mode: 1, Data: 6'
+
+        >>> str(EAMode.parse_ea('(A4)'))
+        'Mode: 2, Data: 4'
+
+        >>> str(EAMode.parse_ea('(A2)+'))
+        'Mode: 3, Data: 2'
+
+        >>> str(EAMode.parse_ea('(A2)-'))  # Invalid, can't do "post-decrement"
+        Traceback (most recent call last):
+        ...
+        AssertionError
+
+        >>> str(EAMode.parse_ea('($45).W'))
+        "Mode: 7, Data: bytearray(b'E')"
+
+        >>> str(EAMode.parse_ea('(%01010111).L'))
+        "Mode: 6, Data: bytearray(b'W')"
+
+        >>> str(EAMode.parse_ea('#$FF'))
+        "Mode: 5, Data: bytearray(b'\\\\xff')"
+
+        >>> str(EAMode.parse_ea('-(A2)'))
+        'Mode: 4, Data: 2'
+        """
+        assert len(addr) >= 2
+
+        if addr[0] == 'D':
+            assert len(addr) == 2
+            assert 0 < int(addr[1]) <= 7
+            return EAMode(EAMode.DRD, int(addr[1]))
+        if addr[0] == 'A':
+            assert len(addr) == 2
+            assert 0 < int(addr[1]) <= 7
+            return EAMode(EAMode.ARD, int(addr[1]))
+        if addr[0] == '(':  # ARI, ARIPI, ALA, or AWA
+            # Parse the inside of the parentheses
+            nested = ""
+            found_paren = False
+
+            i = 1
+            while i < len(addr):
+                if addr[i] == ')':
+                    found_paren = True
+                    break
+
+                nested += addr[i]
+                i += 1
+
+            assert found_paren
+
+            if addr[1] == 'A':  # ARI or ARIPI
+                assert nested[0] == 'A'
+                assert nested[1].isnumeric()
+                assert 0 < int(nested[1]) <= 7
```

Comment:
So the input of 'AA' shoudl throw an AssertionError here?

Reply:
Yup, that should be what happens.